crypto/src/base58/encodable.rs

```diff
@@ -1,122 +1,108 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0
 
-use bin_prot::{from_reader, to_writer};
-pub use bs58::decode::Error;
+use super::*;
 use bs58::encode::EncodeBuilder;
-use serde::{Deserialize, Serialize};
 
 pub trait Base58Encodable {
     /// This is the only part a custom implementation need provide.
     /// Should be a constant from the base58_version_bytes.rs file corresponding
     /// to the type.
     const VERSION_BYTE: u8;
-
-    fn to_base58(&self) -> EncodeBuilder<'static, Vec<u8>>
-    where
-        Self: Sized + Serialize,
-    {
-        let mut buf = Vec::<u8>::new();
-        to_writer(&mut buf, self).unwrap();
-        bs58::encode(buf).with_check_version(Self::VERSION_BYTE)
-    }
-
-    fn to_base58_string(&self) -> String
-    where
-        Self: Sized + Serialize,
-    {
-        self.to_base58().into_string()
-    }
-
-    fn from_base58<'a, I>(i: I) -> Result<Self, bin_prot::error::Error>
-    where
-        I: AsRef<[u8]>,
-        Self: Sized + Deserialize<'a>,
-    {
-        let bytes: Vec<u8> = bs58::decode(i)
-            .with_check(Some(Self::VERSION_BYTE))
-            .into_vec()
-            .map_err(|e| bin_prot::error::Error::Custom {
-                message: format!("{:?}", e),
-            })?;
-
-        // skip the first byte as this still contains the version byte
-        from_reader(&bytes[1..])
-    }
-}
-
-pub trait Base58EncodableHash {
-    /// This is the only part a custom implementation need provide.
-    /// Should be a constant from the base58_version_bytes.rs file corresponding
-    /// to the type.
-    const VERSION_BYTE: u8;
     const MINA_VERSION_BYTE_COUNT: usize = 1;
     const MINA_VERSION_BYTE: u8 = 1;
-    fn to_base58(&self) -> EncodeBuilder<'static, Vec<u8>>
-    where
-        Self: AsRef<[u8; 32]>,
-    {
+    fn to_base58(&self) -> EncodeBuilder<'static, Vec<u8>> {
```

Comment:
I feel strongly that `Base58Encodable` and `Base64Encodable` should have an identical interface since they only differ in the encoding of the output not how the byte representation is accessed

Reply:
In that case, the 2 traits are kind of exclusive to each other, ppl will encounter fn naming conflict if they try to use both traits in the same file, and it will it hard for ppl who read the code later to know exactly which encoding it's doing without looking up in the context

Reply:
Sorry my comment didn't really communicate what I meant. They should use the same process for accessing the underlying bytes. It feels strange for me for `Base64Encodable` to require an `AsRef` implementation to get the bytes and for `Base58Encodable` to require implementing a function to get the bytes. Either is fine but I think they should be the same.

Definitely the public trait interface should be different e.g. `to_base58` and `to_base64`

Reply:
Hmm the `AsRef<[u8]>` approach only works if the structure can provide the bytes as slice of contiguous memory it owns. Maybe the `get_bytes` approach is more general

Reply:
The core difference is that, for mina base58 encoding impls, we have 2 different scenarios that the struct may or may not own the contiguous memory, for either case, we need to deal with the extra mina version bytes hence there has to be owned Vec<u8> as intermediate mem holder.   But for mina base64 encoding case, the owned intermediate mem holder is not needed in any cases, it will just add unnecessary overhead(heap allocation) to create an owned intermediate mem holder if we want to unify the way how to get the encodable bytes.  I've renamed the trait fn in Base58Encodable to make it more clear.  Please note that these 2 encodable trait is only designed to be mina specific and I don't feel that we really want to unify the API at the cost of  extra performance overhead, does that make sense?

Reply:
It makes sense but still doesn't feel right to me. In the interest of getting this PR moving I'll accept what we have and we can maybe revisit it again later.